### **7. Generalized Random Forest**

Developed by Susan Athey, Julie Tibshirani and Stefan Wager, Generalized Random Forest [8] aims to give the solution to a set of local moment equations:
\begin{equation}
  \mathbb{E}\big[\psi_{\tau(s),\nu(s)}(O_i)\big| S_i=s\big]=0,
\end{equation}
where $\tau(s)$ is the parameter we care about and $\nu(s)$ is an optional nuisance parameter. In the problem of Heterogeneous Treatment Effect Evaluation, our parameter of interest $\tau(s)=\xi\cdot \beta(s)$ is identified by 
\begin{equation}
  \psi_{\beta(s),\nu(s)}(R_i,A_i)=(R_i-\beta(s)\cdot A_i-c(s))(1 \quad A_i^T)^T.
\end{equation}
The induced estimator $\hat{\tau}(s)$ for $\tau(s)$ can thus be solved by
\begin{equation}
  \hat{\tau}(s)=\xi^T\left(\sum_{i=1}^n \alpha_i(s)\big(A_i-\bar{A}_\alpha\big)^{\otimes 2}\right)^{-1}\sum_{i=1}^n \alpha_i(s)\big(A_i-\bar{A}_\alpha\big)\big(R_i-\bar{R}_\alpha\big),
\end{equation}
where $\bar{A}_\alpha=\sum \alpha_i(s)A_i$ and $\bar{R}_\alpha=\sum \alpha_i(s)R_i$, and we write $v^{\otimes 2}=vv^T$.

Notice that this formula is just a weighted version of R-learner introduced above. However, instead of using ordinary kernel weighting functions that are prone to a strong curse of dimensionality, GRF uses an adaptive weighting function $\alpha_i(s)$ derived from a forest designed to express heterogeneity in the specified quantity of interest. 
    
To be more specific, in order to obtain $\alpha_i(s)$, GRF first grows a set of $B$ trees indexed by $1,\dots,B$. Then for each such tree, define $L_b(s)$ as the set of training samples falling in the same ``leaf" as x. The weights $\alpha_i(s)$ then capture the frequency with which the $i$-th training example falls into the same leaf as $s$:
\begin{equation}
  \alpha_{bi}(s)=\frac{\boldsymbol{1}\big(\{S_i\in L_b(s)\}\big)}{\big|L_b(s)\big|},\quad \alpha_i(s)=\frac{1}{B}\sum_{b=1}^B \alpha_{bi}(s).
\end{equation}

To sum up, GRF aims to leverage the splitting result of a series of trees to decide the ``localized” weight for HTE estimation at each point $x_0$. Compared with kernel functions, we may expect tree-based weights to be more flexible and better performed in real settings.



In [1]:
# import related packages
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt;
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import LogisticRegression 
from causaldm.learners.Causal_Effect_Learning.Single_Stage.LpRlearner import LpRlearner

ModuleNotFoundError: No module named 'causaldm.learners.Causal_Effect_Learning'

### MovieLens Data

In [2]:
# Get the MovieLens data
import os
os.chdir('/Users/alinaxu/Documents/CDM/CausalDM')
MovieLens_CEL = pd.read_csv("./causaldm/data/MovieLens_CEL.csv")
MovieLens_CEL.pop(MovieLens_CEL.columns[0])
MovieLens_CEL

user_id  movie_id  rating   age  Comedy  Drama  Action  Thriller  \
0         48.0    1193.0     4.0  25.0     0.0    1.0     0.0       0.0   
1         48.0     919.0     4.0  25.0     0.0    1.0     0.0       0.0   
2         48.0     527.0     5.0  25.0     0.0    1.0     0.0       0.0   
3         48.0    1721.0     4.0  25.0     0.0    1.0     0.0       0.0   
4         48.0     150.0     4.0  25.0     0.0    1.0     0.0       0.0   
...        ...       ...     ...   ...     ...    ...     ...       ...   
65637   5878.0    3300.0     2.0  25.0     0.0    0.0     0.0       0.0   
65638   5878.0    1391.0     1.0  25.0     0.0    0.0     0.0       0.0   
65639   5878.0     185.0     4.0  25.0     0.0    0.0     0.0       0.0   
65640   5878.0    2232.0     1.0  25.0     0.0    0.0     0.0       0.0   
65641   5878.0     426.0     3.0  25.0     0.0    0.0     0.0       0.0   

       Sci-Fi  gender_M  occupation_academic/educator  \
0         0.0       1.0                           0.0   
1         0.0       1.0                           0.0   
2         0.0       1.0                           0.0   
3         0.0       1.0                           0.0   
4         0.0       1.0                           0.0   
...       ...       ...                           ...   
65637     1.0       0.0                           0.0   
65638     1.0       0.0                           0.0   
65639     1.0       0.0                           0.0   
65640     1.0       0.0                           0.0   
65641     1.0       0.0                           0.0   

       occupation_college/grad student  occupation_executive/managerial  \
0                                  1.0                              0.0   
1                                  1.0                              0.0   
2                                  1.0                              0.0   
3                                  1.0                              0.0   
4                                  1.0                              0.0   
...                                ...                              ...   
65637                              0.0                              0.0   
65638                              0.0                              0.0   
65639                              0.0                              0.0   
65640                              0.0                              0.0   
65641                              0.0                              0.0   

       occupation_other  occupation_technician/engineer  
0                   0.0                             0.0  
1                   0.0                             0.0  
2                   0.0                             0.0  
3                   0.0                             0.0  
4                   0.0                             0.0  
...                 ...                             ...  
65637               1.0                             0.0  
65638               1.0                             0.0  
65639               1.0                             0.0  
65640               1.0                             0.0  
65641               1.0                             0.0  

[65642 rows x 15 columns]

In [3]:
n = len(MovieLens_CEL)
userinfo_index = np.array([3,5,6,7,8,9,10])
SandA = MovieLens_CEL.iloc[:, np.array([3,4,5,6,7,8,9,10])]

In [4]:
# import related packages
from causaldm._util_causaldm import *

The generalized random forest (GRF) approach has been implemented in package *grf* for R and C++, and *econml* in python. Here we implement the package of *econml* for a simple illustration.

In [5]:
# import the package for Causal Random Forest
! pip install econml

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.0/1.0 MB 2.6 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 433.3/433.3 kB 3.3 MB/s eta 0:00:0000:0100:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 81.0/81.0 kB 2.2 MB/s eta 0:00:00


In [7]:
# A demo code of Causal Random Forest
from econml.grf import CausalForest, CausalIVForest, RegressionForest
from econml.dml import CausalForestDML
est = CausalForest(criterion='het', n_estimators=400, min_samples_leaf=5, max_depth=None,
                    min_var_fraction_leaf=None, min_var_leaf_on_val=True,
                    min_impurity_decrease = 0.0, max_samples=0.45, min_balancedness_tol=.45,
                    warm_start=False, inference=True, fit_intercept=True, subforest_size=4,
                    honest=True, verbose=0, n_jobs=-1, random_state=1235)


est.fit(MovieLens_CEL.iloc[:,userinfo_index], MovieLens_CEL['Drama'], MovieLens_CEL['rating'])

HTE_GRF = est.predict(MovieLens_CEL.iloc[:,userinfo_index], interval=False, alpha=0.05)
HTE_GRF = HTE_GRF.flatten()


Let's focus on the estimated HTEs for three randomly chosen users:

In [8]:
print("Generalized Random Forest:  ",HTE_GRF[np.array([0,300,900])])

Generalized Random Forest:   [0.3605 0.3783 0.3646]


In [9]:
ATE_GRF = np.sum(HTE_GRF)/n
print("Choosing Drama instead of Sci-Fi is expected to improve the rating of all users by",round(ATE_GRF,4), "out of 5 points.")

Choosing Drama instead of Sci-Fi is expected to improve the rating of all users by 0.358 out of 5 points.


**Conclusion:** Choosing Drama instead of Sci-Fi is expected to improve the rating of all users by 0.358 out of 5 points.

## References

8. Susan Athey, Julie Tibshirani, and Stefan Wager. Generalized random forests. The Annals of Statistics, 47(2):1148–1178, 2019.